src/dex/ekubo/ekubo.ts

```diff
@@ -0,0 +1,657 @@
+    }),
+  ),
+});
+
+const allPoolsSchema = Joi.array<
+  {
+    core_address: string;
+    token0: string;
+    token1: string;
+    fee: string;
+    tick_spacing: number;
+    extension: string;
+  }[]
+>().items(
+  Joi.object({
+    core_address: Joi.string(),
+    token0: Joi.string(),
+    token1: Joi.string(),
+    fee: Joi.string(),
+    tick_spacing: Joi.number(),
+    extension: Joi.string(),
+  }),
+);
+
+const MIN_TICK_SPACINGS_PER_POOL = 2;
+const MAX_BATCH_SIZE = 100;
+
+const POOL_MAP_UPDATE_INTERVAL_MS = 5 * 60 * 1000;
+
+// Ekubo Protocol https://ekubo.org/
+export class Ekubo extends SimpleExchange implements IDex<EkuboData> {
+  public readonly hasConstantPriceLargeAmounts = false;
+  public readonly needWrapNative = false;
+  public readonly isFeeOnTransferSupported = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(EkuboConfig);
+
+  private readonly pools: Map<string, BasePool> = new Map();
+
+  public logger;
+
+  public readonly config;
+  public readonly routerIface;
+
+  private readonly core;
+  private readonly coreIface;
+  private readonly dataFetcher;
+  private readonly supportedExtensions;
+
+  private interval?: NodeJS.Timeout;
+
+  // Caches the number of decimals for TVL computation purposes
+  private readonly decimals: Record<string, number> = {
+    [ETHER_ADDRESS]: 18,
+  };
+
+  public constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+  ) {
+    super(dexHelper, dexKey);
+
+    this.logger = dexHelper.getLogger(dexKey);
+    this.config = EkuboConfig[dexKey][network];
+    this.core = new Contract(this.config.core, CoreABI, dexHelper.provider);
+    this.coreIface = new Interface(CoreABI);
+    this.dataFetcher = new Contract(
+      this.config.dataFetcher,
+      DataFetcherABI,
+      dexHelper.provider,
+    );
+    this.routerIface = new Interface(RouterABI);
+
+    // 0 are vanilla pools
+    this.supportedExtensions = [0n, BigInt(this.config.oracle)];
+  }
+
+  // Periodically schedules fetching pool keys from the Ekubo API and filling in details with the quote data fetcher
+  public async initializePricing(blockNumber: number) {
+    await this.updatePoolMap(blockNumber);
+
+    this.interval = setInterval(async () => {
+      await this.updatePoolMap(await this.dexHelper.provider.getBlockNumber());
+    }, POOL_MAP_UPDATE_INTERVAL_MS);
+  }
+
+  // LEGACY
+  public getAdapters(
+    _side: SwapSide,
+  ): { name: string; index: number }[] | null {
+    return null;
+  }
+
+  public async getPoolIdentifiers(
+    srcToken: Token,
+    destToken: Token,
+    _side: SwapSide,
+    blockNumber: number,
+  ): Promise<string[]> {
+    const [token0, token1] = sortAndConvertTokens(srcToken, destToken);
+    const pair = hexStringTokenPair(token0, token1);
+
+    let poolKeys: PoolKey[];
+    try {
+      poolKeys = (await this.fetchAllPoolKeys()).filter(
```

Comment:
We cannot make this http request each time, in this implementation it should be cached and refreshed with some interval then